blockchain/blockindex.go

```diff
@@ -574,14 +883,424 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
+	// an invalid ancestor as follows:
+	//
+	// Tip 105a: 105a, 104a, 103a       (102a is failed block, next)
+	// Tip 106b: 106b, 105b, 104a, 103a (102a is failed block, next)
+	// Tip 106c: 106c, 105c, 104c, 103a (102a is failed block, next)
+	// Tip 105d: 105d, 104d, 103a       (102a is failed block, next)
+	//
+	// Note that it might be tempting to consider trying to optimize this to
+	// skip directly to the next tip once a node with a common invalid ancestor
+	// is found.  However, that would result in incorrect behavior if a block is
+	// marked invalid deeper in a branch first and then an earlier block is
+	// later marked invalid as it would result in skipping the intermediate
+	// blocks thereby NOT marking them as invalid as they should be.
+	//
+	// For example, consider what happens if 105c is marked invalid prior to the
+	// block data for 102a becoming available and found to be invalid.  Blocks
+	// 106c and 105c would already be marked invalid, and blocks 104c and 103a
+	// need to be marked invalid.
+	markDescendantsInvalid := func(node, tip *blockNode) {
+		// Nothing to do if the node is not an ancestor of the given chain tip.
+		if tip.Ancestor(node.height) != node {
+			return
+		}
+
+		// Set this chain tip as the best known invalid block when it has more
+		// work than the current one.
+		bi.maybeUpdateBestInvalid(tip)
+
+		// Mark everything that descends from the failed block as having an
+		// invalid ancestor.
+		for n := tip; n != node; n = n.parent {
+			// Skip blocks that are already known to have an invalid ancestor.
+			if n.status.KnownInvalidAncestor() {
+				continue
+			}
+
+			bi.setStatusFlags(n, statusInvalidAncestor)
+			bi.unsetStatusFlags(n, statusValidated)
+			bi.removeBestChainCandidate(n)
+
+			// Remove any children that depend on the failed block from the set
+			// of unlinked blocks accordingly since they are no longer eligible
+			// for connection even if the full block data for a block becomes
+			// available.
+			delete(bi.unlinkedChildrenOf, n)
+		}
+	}
+
+	// Chain tips at the same or lower heights than the failed block can't
+	// possibly be descendants of it, so use it as the lower height bound filter
+	// when iterating chain tips.  Note that this will make use of the cache of
+	// recent tips when possible.
+	bi.forEachChainTipAfterHeight(node, func(tip *blockNode) error {
+		markDescendantsInvalid(node, tip)
+		return nil
+	})
+
+	// Update the best header if the current one is now invalid which will be
+	// the case when the best header is a descendant of the failed block.
+	if bi.bestHeader.status.KnownInvalid() {
+		// Use the first ancestor of the failed block that is not known to be
+		// invalid as the lower bound for the best header.  This will typically
+		// be the parent of the failed block, but it might be some more distant
+		// ancestor when performing manual invalidation.
+		n := node.parent
+		for n != nil && n.status.KnownInvalid() {
```

Comment:
Maybe you also need to include `&& n.status.KnownInvalidAncestor()` here as well? Otherwise performing a manual invalidation might make a header with an invalid parent the best header, unless that is protected at the entrypoint for manual invalidation.

Reply:
`KnownInvalid` includes the invalid ancestor check.

```Go
func (status blockStatus) KnownInvalid() bool {
	return status&(statusValidateFailed|statusInvalidAncestor) != 0
}
```

Reply:
Bleh... :P